src/TRoom.cpp

```diff
@@ -35,6 +35,33 @@
 #include <QStringBuilder>
 #include "post_guard.h"
 
+// Helper needed to allow Qt::PenStyle enum to be serialised - the compilation
+// errors that result in not having this are really confusing...!
+// Curiously the reverse (deserialising) when reading them (as part of the
+// QMap<QString, Qt::PenStyle> customLinesStyle SEEMS possible without any
+// "help" but WILL ACTUALLY CORRUPT THE READ OF THE data stream as instead of
+// writing a single byte (?) as a "quint8" a whole chunk of bytes will be
+// writen out as a  full multi-byte "int" value which will totally mess up
+// the data afterwards...)
+QDataStream &operator>>(QDataStream& ds, Qt::PenStyle& value)
+{
+    quint8 temporary;
+    ds >> temporary;
+    switch(temporary) {
```

Comment:
Wouldn't it be a lot simplier to initialise it to `Qt::SolidLine` and have an `if` for `Qt::DashDotDotLine`?

Reply:
Allows a fall-through to a meaningful/usable value should something else show up (there are a couple of other `Qt::PenStyles` {none and custom} which we cannot usefully use...

Reply:
Okay, but see my comment - just an initialisation and an _if_ will take a lot less space, and it won't give warnings on GCC.

Reply:
Seems like there ought to be a Qt provided MACRO for this that can drop in an appropriate `[[`..`fallthrough]]` "attribute" which GCC7 is to get alongside Clang which has had it's own for a while now...

Reply:
I believe that got added to C++17, but we unfortunately can't use it yet because Linux distros are keeping us down. That said we can fine our own fallthrough macro right now that'll work: https://infektor.net/posts/2017-01-19-using-cpp17-attributes-today.html#using-the-fallthrough-attribute

Reply:
Noted, though where would we define it? It is hard to code a `#define` macro that takes _arguments_ in the QMake `mudlet.pro` file IIRC...

Reply:
The point I'm making here is that we don't need this complicated switch :smile: we can simply do:

```
value = Qt::SolidLine;
if (Qt::DashDotDotLine) {
   value = static_cast<Qt::PenStyle>(temporary);
}
```

3 lines of code compared to 12

Reply:
That code doesn't make sense - the `switch` is being used to validate 4 specific values of `temporary` {the, cast to a `quint8`, values of `Qt::DotLine`, `Qt::DashLine`, `Qt::DashDotLine`, `Qt::DashDotDotLine`} and to use which ever of __them__ if they are found plus a fifth value which is used __for all other (252) cases__ (including the remaining valid one of `Qt::SolidLine`)...

Going for an `if` based it would be:

    if ( static_cast<Qt::PenStyle>(temporary) == Qt::DotLine || static_cast<Qt::PenStyle>(temporary) == Qt::DashLine || static_cast<Qt::PenStyle>(temporary) == Qt::DashDotLine || static_cast<Qt::PenStyle>(temporary) == Qt::DashDotDotLine ) {
        value = static_cast<Qt::PenStyle>(temporary);
    } else {
        value = Qt::SolidLine;
    }

In this situation I actually find the `switch` one clearer. 😲

(It is only the need to tell suspicious compliers that the running together of those `case`s is deliberate and wanted which requires three of those lines - I could swap the `[[clang::fallthrough]]` for `; // Fall-through` though and that should shut each one of them up!)

Reply:
OK, how about the following?

```c++
    switch (temporary) {
    case Qt::DotLine | Qt::DashLine | Qt::DashDotLine:
        [[clang::fallthrough]];
    case Qt::DashDotDotLine:
        value = static_cast<Qt::PenStyle>(temporary);
        break;
    default:
        // Force anything else to be a solidline
        value = Qt::SolidLine;
    }
```

Reply:
Is that a new-fangled C++11 thing, being able to OR `case` value together? Unfortunately I do not think the individual values are 2^n values so they can't be used in the manner of a bitmask either...

Reply:
No, it's something that was around for a while: https://doc.qt.io/qt-5/qflags.html#operator-7c-eq

Reply:
Looking at the current definition of that enum:
> Constant | Value | Description
> -- | -- | --
> Qt::NoPen | 0 | no line at all. For example, QPainter::drawRect() fills but does not draw any boundary line.
> Qt::SolidLine | 1 | A plain line.
> Qt::DashLine | 2 | Dashes separated by a few pixels.
> Qt::DotLine | 3 | Dots separated by a few pixels.
> Qt::DashDotLine | 4 | Alternate dots and dashes.
> Qt::DashDotDotLine | 5 | One dash, two dots, one dash, two dots.
> Qt::CustomDashLine | 6 | A custom pattern defined using QPainterPathStroker::setDashPattern().

A bitwise-OR doesn't work to provide a `case` value that can be used to detect `Qt::DotLine` OR `Qt::DashLine` OR `Qt::DashDotLine` unless C++11 has gained something that I am not aware of..

Reply:
Oh. Nevermind me then!